### ISSUE TITLE
Is there any good solution for making fullpage screenshots on lazy load pages?

### ISSUE THREAD

**Issue by ghost** [microsoft/playwright#19861]
I was wondering if there is any good solution out there to this problem. And I'm searching for a more generic solution for making fullpage screenshots on different websites with the lazy loaded content loaded. especially images.

The only thing I had in mind was something like scrolling to the bottom of the page or scrollIntoViewIfNeeded with a bottom element. But this doesn't work so far.

**Comment by cstickel** [comment]
@karlhorky be aware that `scrollIntoViewIfNeeded` might mess up scroll positions of scrollable elements. like for example if you do have a horizontal slider with images it would not be in the inital position anymore but slided that the last element is visible.

the solution to resize the viewport to the scrollheight, that everything is in viewport and then waiting for the network to be idle and then resize it back might work for in those cases.

so still there is imho no perfect solution for that problem yet.

**Comment by zarart99** [comment]
We encountered a similar challenge with lazy-loaded images during visual regression testing. Our solution involves programmatically forcing the lazy-loaded images to load by modifying the loading attribute and resetting the src. Here’s our current approach:
```
async function forceLoadLazyImages(page: Page): Promise<void> {
  await page.evaluate(() => {
    document.querySelectorAll('img[loading="lazy"]').forEach((img) => {
      const image = img as HTMLImageElement;
      image.setAttribute('loading', 'eager'); // Force eager loading
      const src = image.src;
      image.src = '';  // Reset src to reload the image
      image.src = src; // Set src back to original
    });
  });
}
```

**Comment by tremby** [comment]
@zarart99 that was helpful to me today; thank you.

I found that I didn't need to unset and reset the src attribute (maybe because I'm using srcsets? ymmv), and just in case you didn't know, you can hint to qSA what type to expect:

```typescript
async function forceLoadLazyImages(page: Page): Promise<void> {
  return page.evaluate(() => {
    for (const image of document.querySelectorAll<HTMLImageElement>('img[loading="lazy"]')) {
      image.setAttribute('loading', 'eager');
    }
  });
}
```

**Comment by ahayes91** [comment]
I went deep into the rabbit hole today using this issue and https://github.com/microsoft/playwright/issues/620, as I was seeing some serious flakiness with `page.screenshot({fullPage: true})` in Chromium. I won't share any proprietary code or screenshots, but will do my best to explain below. 

**TL;DR, switching to Firefox solved the intermittent image loading failures for us!**

## Problem - images below the fold were inconsistently loading in fullPage Chromium screenshots

I was seeing intermittent failures where lazy loaded images "below the fold" weren't appearing in my screenshots. This was persistent, no matter how much full page scrolling (through https://github.com/microsoft/playwright/issues/620#issuecomment-578022596) or waiting for individual DOM elements through `page.locator().waitFor()`, or any [waitForResponse](https://playwright.dev/docs/api/class-page#page-wait-for-response) on the image sources themselves, or even hard `waitForTimeout` I did prior to taking the screenshot.

## Root cause (probably) - `captureBeyondViewport` Chrome DevTools Protocol doesn't trigger viewport events

After a bunch of debugging on what was going on, I tracked it down to the `fullPage` attribute specifically. I removed `fullPage: true`, and continued doing all the same stuff I did before (scrolling the fullPage and waiting for locators and waiting for responses), and I scrolled the viewport to get a little bit of one of the images into view through `await page.evaluate(() => window.scrollTo(0, 100));`. Without `fullPage`, the partial screenshot produced by `page.screenshot()` was able to show me that the images were indeed loading.

For context, I'm in the process of transitioning away from Cypress screenshot (using Chrome) to Playwright screenshot. 
Cypress screenshot, when using `fullPage`, scrolls and takes multiple screenshots and stitches them together, as per the docs at https://docs.cypress.io/api/commands/screenshot#Full-page-captures-and-fixedsticky-elements. 

> When passing `fullPage` to the `capture` option, Cypress scrolls the application under test from top to bottom, takes screenshots at each point and stitches them together. 

I had a suspicion that maybe Playwright screenshot didn't work the same way, so I pulled a fork of this repo and asked Cursor to give me a synopsis, which seems reasonable to me:

> The Problem
> When fullPage: true is used for screenshots in Chromium-based browsers, Playwright uses Chrome DevTools Protocol's Page.captureScreenshot with captureBeyondViewport: true. This captures the entire page without actually scrolling it.
> Here's the key code in /code/playwright/packages/playwright-core/src/server/chromium/crPage.ts:
> crPage.ts Lines 253-274

```
async takeScreenshot(progress: Progress, format: 'png' | 'jpeg', documentRect: types.Rect | undefined, viewportRect: types.Rect | undefined, quality: number | undefined, fitsViewport: boolean, scale: 'css' | 'device'): Promise<Buffer> {
  const { visualViewport } = await progress.race(this._mainFrameSession._client.send('Page.getLayoutMetrics'));
  if (!documentRect) {
    documentRect = {
      x: visualViewport.pageX + viewportRect!.x,
      y: visualViewport.pageY + viewportRect!.y,
      ...helper.enclosingIntSize({
        width: viewportRect!.width / visualViewport.scale,
        height: viewportRect!.height / visualViewport.scale,
      })
    };
  }
  // When taking screenshots with documentRect (based on the page content, not viewport),
  // ignore current page scale.
  const clip = { ...documentRect, scale: viewportRect ? visualViewport.scale : 1 };
  if (scale === 'css') {
    const deviceScaleFactor = this._browserContext._options.deviceScaleFactor || 1;
    clip.scale /= deviceScaleFactor;
  }
  const result = await progress.race(this._mainFrameSession._client.send('Page.captureScreenshot', { format, quality, clip, captureBeyondViewport: !fitsViewport }));
  return Buffer.from(result.data, 'base64');
}
```

> And in /code/playwright/packages/playwright-core/src/server/screenshotter.ts:
> screenshotter.tsLines 210-216

```
if (options.fullPage) {
  const fullPageSize = await this._fullPageSize(progress);
  let documentRect = { x: 0, y: 0, width: fullPageSize.width, height: fullPageSize.height };
  const fitsViewport = fullPageSize.width <= viewportSize.width && fullPageSize.height <= viewportSize.height;
  if (options.clip)
    documentRect = trimClipToSize(options.clip, documentRect);
  return await this._screenshot(progress, format, documentRect, undefined, fitsViewport, options);
```

> Why Images Don't Load

> When captureBeyondViewport: true is used:
> 1. The page never actually scrolls - Chrome captures the entire rendering surface without moving the viewport
> 2. Lazy-loaded images don't trigger - Images with:
>   - loading="lazy" attribute
>   - JavaScript-based lazy loading (Intersection Observer, scroll listeners)
>   - Images loaded when they come into view
> ...never receive the trigger to load because they never enter the viewport
> 3. The screenshot captures placeholder content - Instead of the actual images, you get blank spaces, spinners, or placeholder icons

> This is a fundamental limitation of how Chrome's captureBeyondViewport works - it's optimized for speed and doesn't trigger viewport-based events. This is by design in the Chrome DevTools Protocol, not a bug in Playwright per se.

## Solution - move to Firefox!

I moved to Firefox and stopped seeing these intermittent issues. Great!

But why? I turned to Cursor again, and asked it to assess how Playwright's screenshot methodology is different in Firefox, and here's the kicker - Cursor initially told me it thought it would be likely that Firefox AND Webkit would see the same issue. I nudged it again, saying that I was seeing the issue as resolved, and it did some soul searching before coming up with the following answer:

> The critical difference is in how the rendering engines handle off-screen content:
> Chromium's Approach (captureBeyondViewport: true)
> Looking at line 272 in crPage.ts:
> crPage.tsLine 272
```
const result = await progress.race(this._mainFrameSession._client.send('Page.captureScreenshot', { format, quality, clip, captureBeyondViewport: !fitsViewport }));
```

> When Chromium captures with captureBeyondViewport: true:
> 1. It captures from the existing compositor surface/backing store
> 2. This is optimized for performance - it doesn't force a new render pass
> 3. Images with loading="lazy" or viewport-based lazy loading never receive the signal to load because the viewport never actually moves
> 4. The compositor has placeholder/blank content for those images

> Firefox's Approach (drawSnapshot())
> Firefox's drawSnapshot() in the Gecko rendering engine:
> 1. Forces a complete paint operation for the specified rect
> 2. When painting a large rect that includes off-screen content, Gecko calculates which resources are needed
> 3. Firefox's implementation of loading="lazy" might be less strict - it may decide to load images when they're part of a paint operation, even if they're not in the viewport
> 4. Or, Firefox might temporarily suspend lazy-loading behavior during the snapshot operation to ensure complete rendering

> The Evidence
> The web search revealed that there was actually a [Firefox bug (1650677)](https://bugzilla.mozilla.org/show_bug.cgi?id=1650677) related to this, suggesting Firefox developers were aware of and potentially fixed issues with lazy-loaded content in screenshots.

> Practical Difference

```
// Chromium: Captures from existing surface
// "Take what's already painted, even if some images aren't loaded"

// Firefox drawSnapshot: Forces fresh paint
// "Render this rect completely, loading what's needed"
```

> This is why you see:
> - Chromium: Missing lazy-loaded images in fullPage: true screenshots
> - Firefox: Lazy-loaded images appear correctly in fullPage: true screenshots
> The Firefox Gecko engine's drawSnapshot() API fundamentally triggers more of the rendering pipeline than Chromium's surface capture approach, which has the side effect of loading lazy images that would otherwise remain unloaded.

Do I believe the robot's explanation? Maybe - but it seems a bit fluffy to me.

### Would Chrome, rather than Chromium help?

I didn't test this and it's a good question. You can use [Chrome](https://playwright.dev/docs/browsers#google-chrome--microsoft-edge) as a channel, but personally I couldn't get that working in the Playwright [Docker image](https://playwright.dev/docs/docker) I was using (my whole setup is also using [Storybook Test Runner](https://storybook.js.org/docs/writing-tests/integrations/test-runner) just to further complicate things). 
I stuck with Chromium during testing here (and the only thing I had to do to get my existing tests working, albeit with intermittent image loading issues, was to convert some `mp4` files to `webm` using `ffmpeg` - everything else was the same).

### If you absolutely need to use Chromium - is there anything you could do?

I did try masking the images (https://playwright.dev/docs/api/class-page#page-screenshot), which worked to give me consistency - but it was really ugly, and a degradation of the experience we previously had with Cypress and Chrome. Worth noting here though as a possibility for anyone who gets stuck.

## The end

Hopefully this helps someone else who comes across this in the future! If I ever get a chance to make a reproduction repository, I'll raise a new ticket with Playwright about this fullPage and Chromium browser flakiness because it was really weird. I'm sure Playwright maintainers would definitely have more insight than the Cursor robot about the root cause too 😏

**Comment by MeetSnowmaker** [comment]
Hi,

Since this issue is one of the top Google results for _"playwright lazy loading images stable screenshots"_, I wanted to share the approach that finally gave us **340/340 passing** visual regression tests across **10 browser/device profiles** (Chrome, Firefox, WebKit/Safari, 4× mobile, 2× tablet) on a Next.js + MUI static-site monorepo.

We hit **six separate root causes**. Every one of them had to be fixed before runs were reliably stable across 5 consecutive full passes.

---

## Setup

- `@playwright/test` **1.58.2** (Docker image `mcr.microsoft.com/playwright:v1.58.2-jammy`)
- Next.js 16 static export (`output: 'export'`)
- MUI 7 (`sx` prop animations with staggered `animationDelay`)
- Images with `loading="lazy"` and `decoding="async"`
- MUI `ImageList variant="masonry"` (CSS columns grid)
- `fullPage: true` screenshots, 10 browser projects

---

## Root cause 1 — `animation-delay` is not zeroed by `reducedMotion: 'reduce'`

Setting `reducedMotion: 'reduce'` zeroes `animation-duration` but **does not touch `animation-delay`**. MUI gallery/masonry grids use staggered delays like `animationDelay: '${index * 50}ms'`, so items sit at `opacity: 0` during their delay window then jump to `opacity: 1`. Playwright's two consecutive stability frames see different opacities → _"Failed to take two consecutive stable screenshots"_.

**Fix:** inject a stylesheet before any page code runs:

```ts
style.textContent = '*, *::before, *::after { animation-delay: 0ms !important; }';
```

---

## Root cause 2 — `decoding="async"` makes `img.complete` fire before pixels are painted

For `fullPage: true`, the browser expands the virtual viewport which triggers `IntersectionObserver` for off-screen lazy images. They start loading _during_ Playwright's stability check → frames differ.

The standard fix is to switch `loading="lazy"` → `loading="eager"` after `networkidle`, then wait for `img.complete`. **This is not enough.** With `decoding="async"`, `img.complete` becomes `true` once the network fetch finishes — but the browser may not have decoded the pixel data yet. Two stability frames can still capture different rendered states.

**Fix:** use `img.decode()` instead of polling `img.complete`. `HTMLImageElement.decode()` returns a Promise that resolves only when the image is fully decoded and _ready to paint_, regardless of the `decoding` attribute.

---

## Root cause 3 — CSS columns/masonry layout reflows after `img.decode()` resolves

`img.decode()` gives you decoded pixels, but **CSS columns layout** (`ImageList variant="masonry"`) recalculates column assignments _in the next render cycle_ after image dimensions become known. If Playwright's stability check starts before that reflow is painted, one frame shows empty columns and the next shows them filled.

**Fix:** double `requestAnimationFrame` after `img.decode()`:

```ts
// First rAF = layout scheduled, second rAF = layout painted.
await new Promise<void>((r) => requestAnimationFrame(() => requestAnimationFrame(r)));
```

---

## Root cause 4 — `document.fonts.ready` vs font paint cycle

Web fonts finish loading asynchronously. Even after `networkidle`, a font-swap reflow can happen between Playwright's two stability frames if fonts aren't explicitly waited on.

**Fix:** run `document.fonts.ready` in parallel with `img.decode()`:

```ts
await Promise.all([...imgs.map((img) => img.decode().catch(() => {})), document.fonts.ready]);
```

---

## Root cause 5 — Firefox auto-focuses the first interactive element on load

Firefox automatically focuses the first interactive element (a link, button, or input) when a page loads. `:focus-visible` and Firefox's own `:-moz-focusring` (a dotted border that is a **separate CSS primitive from `outline`**) then render on that element. Between two runs, which element gets focused and when can vary → pixel diff on text links.

**Fix:** two layers:

1. Suppress all focus indicators in the init script:

```ts
':focus, :focus-visible, :-moz-focusring { outline: none !important; box-shadow: none !important; }';
```

2. Programmatically blur the active element after decode:

```ts
if (document.activeElement instanceof HTMLElement) {
  document.activeElement.blur();
}
```

---

## Root cause 6 — GPU compositing and sub-pixel font hinting are non-deterministic

Firefox and WebKit on Linux render WebP/JPEG images with ±1 pixel colour variation between frames due to GPU compositing. Linux Firefox sub-pixel font hinting also shifts thin glyphs by a fraction of a pixel depending on render timing. These are not fixable in JavaScript or CSS.

**Fix:** a small `maxDiffPixelRatio` in the Playwright config. On a typical full-page screenshot, 0.2% covers the GPU noise while leaving any real layout or content regression detectable:

```ts
expect: {
  toHaveScreenshot: {
    maxDiffPixelRatio: 0.002,
  },
},
```

---

## Complete solution

### `tests/visual/base.ts`

```ts
import { expect, test as base, type Page } from '@playwright/test';

export const test = base.extend<{ page: Page }>({
  page: async ({ page }, use) => {
    await page.addInitScript(() => {
      const style = document.createElement('style');
      style.textContent = [
        // RC1: zero animation delays (reducedMotion only zeroes duration)
        '*, *::before, *::after { animation-delay: 0ms !important; }',
        // RC5: suppress Firefox auto-focus ring on first interactive element
        ':focus, :focus-visible, :-moz-focusring { outline: none !important; box-shadow: none !important; }',
      ].join('\n');
      document.documentElement.appendChild(style);
    });
    await use(page);
  },
});

export async function waitForImages(page: Page): Promise<void> {
  // RC2 step 1: trigger lazy images that are already in the DOM
  await page.evaluate(() => {
    for (const img of document.querySelectorAll<HTMLImageElement>('img[loading="lazy"]')) {
      img.setAttribute('loading', 'eager');
    }
  });

  await page.evaluate(async () => {
    const imgs = [...document.querySelectorAll<HTMLImageElement>('img')];

    // RC2: img.decode() waits for fetch + decode (img.complete fires too early
    //      for decoding="async" images).
    // RC4: document.fonts.ready prevents font-swap reflow between frames.
    await Promise.all([...imgs.map((img) => img.decode().catch(() => {})), document.fonts.ready]);

    // RC5: blur whatever Firefox auto-focused on page load
    if (document.activeElement instanceof HTMLElement) {
      document.activeElement.blur();
    }

    // RC3: CSS columns/masonry layout reflows AFTER img.decode() resolves.
    // Double-rAF: first frame = layout scheduled, second = layout painted.
    await new Promise<void>((r) => requestAnimationFrame(() => requestAnimationFrame(r)));
  });

  // Catch any secondary requests the image loads may have triggered
  await page.waitForLoadState('networkidle');
}

export { expect };
```

### `playwright.config.ts`

```ts
import { defineConfig, devices } from '@playwright/test';

export default defineConfig({
  testDir: './tests/visual',
  snapshotDir: './tests/snapshots',
  fullyParallel: true,
  forbidOnly: !!process.env.CI,
  retries: process.env.CI ? 1 : 0,
  reporter: [['html'], ['json', { outputFile: 'playwright-report/report.json' }], ['list']],

  use: {
    contextOptions: {
      // Zeroes animation-duration. Does NOT zero animation-delay — base.ts handles that.
      reducedMotion: 'reduce',
    },
  },

  expect: {
    toHaveScreenshot: {
      // RC6: GPU image compositing (WebP/JPEG) and Linux sub-pixel font hinting
      // produce ±1 pixel non-deterministic variation between frames in Firefox
      // and WebKit. 0.2% covers the noise; any real regression affects far more.
      maxDiffPixelRatio: 0.002,
    },
  },

  // Serve your pre-built static sites. Set CI=1 in Docker so reuseExistingServer
  // is false — otherwise Playwright leaves the serve processes alive and the
  // container never exits.
  webServer: [
    {
      command: 'npx serve my-site/out -l 3000',
      port: 3000,
      reuseExistingServer: !process.env.CI,
      timeout: 10_000,
    },
  ],

  projects: [
    // Fast dev profile
    { name: 'desktop-chrome', use: { ...devices['Desktop Chrome'] } },
    { name: 'iphone-14-pro',  use: { ...devices['iPhone 14 Pro'] } },

    // Full PR profile
    { name: 'desktop-firefox',      use: { ...devices['Desktop Firefox'] } },
    { name: 'desktop-safari',       use: { ...devices['Desktop Safari'] } },
    { name: 'iphone-14-pro-max',    use: { ...devices['iPhone 14 Pro Max'] } },
    { name: 'iphone-se',            use: { ...devices['iPhone SE'] } },
    { name: 'pixel-7',              use: { ...devices['Pixel 7'] } },
    { name: 'galaxy-s8',            use: { ...devices['Galaxy S8'] } },
    { name: 'ipad-pro-landscape',   use: { ...devices['iPad Pro 11 landscape'] } },
    { name: 'ipad-mini',            use: { ...devices['iPad Mini'] } },
  ],
});
```

### `tests/visual/example.spec.ts`

```ts
import { expect, test, waitForImages } from './base';

test.use({ baseURL: 'http://localhost:3000' });

const pages = [
  { name: 'home', path: '/' },
  { name: 'gallery', path: '/gallery/' },
  { name: 'portfolio', path: '/portfolio/' },
  { name: 'contact', path: '/contact/' },
];

for (const { name, path } of pages) {
  test(name, async ({ page }) => {
    await page.goto(path);
    await page.waitForLoadState('networkidle');
    await waitForImages(page);
    await expect(page).toHaveScreenshot(`${name}.png`, { fullPage: true });
  });
}
```

### Bonus: if your pages show a snackbar on first visit

Use `context.addInitScript` in `test.beforeEach`, **not** `page.addInitScript` inside the test body. Context-level scripts are registered before the page exists and are guaranteed to run before any React code:

```ts
test.beforeEach(async ({ context }) => {
  await context.addInitScript(() => {
    sessionStorage.setItem('demo-notice-shown', '1');
  });
});
```

---

Hope this helps someone.

_Yeah, I used Opus to make sense of all the suffering we endured and compiled it into an md._

**Comment by webmox** [comment]
One thing I haven't seen mentioned here — IntersectionObserver-based lazy loading behaves differently depending on how you scroll.

If you jump straight to document.body.scrollHeight in one call, observers for elements in the middle of the page may never fire. The browser doesn't synthesize intersection events for elements that were "passed through" — only for elements currently entering or leaving the viewport.

Incremental scrolling fixes this:


```js 
async function triggerLazyContent(page) {
  const viewportHeight = page.viewportSize().height;
  let scrollY = 0;
  
  while (true) {
    scrollY += Math.floor(viewportHeight * 0.8); // 80% overlap
    await page.evaluate(y => window.scrollTo(0, y), scrollY);
    await page.waitForTimeout(300);
    
    const atBottom = await page.evaluate(
      () => window.scrollY + window.innerHeight >= document.body.scrollHeight - 10
    );
    if (atBottom) break;
  }
  
  // scroll back to top before full-page capture
  await page.evaluate(() => window.scrollTo(0, 0));
  await page.waitForTimeout(500);
}
```
The 80% viewport overlap ensures every element spends enough time "in view" for observers to fire. waitForTimeout(300) gives images time to start loading after the observer triggers.

Also worth noting: loading="lazy" on <img> tags and IntersectionObserver-based lazy loading are two different mechanisms. The img[loading="lazy"] approach from @karlhorky works for native lazy loading, but many sites use custom JS observers (often through libraries like lozad.js or lazysizes). For those, scrolling is the only reliable trigger.

For sites using content-visibility: auto (increasingly common in 2026), you need to scroll AND wait for contentvisibilityautostatechange events, or just add a longer delay after each scroll step.
